google/cloud/storage/status_or.h

```diff
@@ -0,0 +1,178 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_STORAGE_STATUS_OR_H_
+#define GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_STORAGE_STATUS_OR_H_
+
+#include "google/cloud/internal/throw_delegate.h"
+#include "google/cloud/storage/internal/throw_status_delegate.h"
+#include "google/cloud/storage/status.h"
+#include <type_traits>
+#include <utility>
+
+namespace google {
+namespace cloud {
+namespace storage {
+inline namespace STORAGE_CLIENT_NS {
+/**
+ * `StatusOr<T>` is used to return a value or an error status when exceptions
+ * are disabled.
+ *
+ * If the library is compiled with exceptions disabled there is no standard C++
+ * mechanism to report errors. In that case we use this class to wrap the values
+ * returned to the application.
+ *
+ * The application typically calls an API in the library and must check the
+ * returned error status:
+ *
+ * @code
+ * namespace gcs = google::cloud::storage;
+ * void AppCode(gcs::noex::Client client) {
```

Comment:
One of the things that makes `StatusOr<T>` and its spiritual sibling, `optional<T>`, convenient to use is their pointer-like interface and syntax. That is, defining and using operator*() and operator->(). I realize this is a first cut of a minimal class, but I think we may want to add these operators now. Additionally, we may wan to use them in documentation as a way of highlighting and somewhat encouraging them.

The similarities are:

explicit operator bool is the same as .ok().
.value() is similar to operator*(), EXCEPT that .value() is defined to throw/abort whereas op*() is UB if there's no value. operator->() is also UB if there's no value.

So in this example here where the error is checked, I'd suggest that we use the pointer interface because it's natural, less typing, and potentially more efficient:

```cc
void AppCode(gcs::noex::Client client) {
  gcs::StatusOr<gcs::BucketMetadata> result = client.GetBucketMetadata("foo");
  if (!result) {
    std::cerr << result.status() << '\n';
    return;
  }
  const gcs::BucketMetadata& meta = *result;
  ...
}
```

We could then show another example where the above code is shortened to a simple call to .value(), which is defined to throw/abort, and avoid the explicit error checking.

... thoughts?

Reply:
Done.  I think your example implies an `operator bool()` so I added that too. I made it explicit but have no strong opinion on that.

Reply:
Yep. LGTM. We do want explicit on that so:

```cc
bool b = sor;  // Won't compile
if (sor) ...; // Will compile
```